### PR TITLE
Add minimum timeouts to MWA spec

### LINKED ISSUE #51: Spec specifies X for timeouts
Determine appropriate timeouts for X placeholders in the spec

### PR DESCRIPTION
**Pull request by sdlaver**
Fixes #51

### CODE REVIEW

File: spec/spec.md [hunk 1]
```diff
@@ -132,11 +132,11 @@ where:
 
 Once the URI is opened, the dapp endpoint should attempt to connect to the local WebSocket address, `ws://localhost:<port_number>/solana-wallet`, and proceed to [Session establishment](#session-establishment).
 
-If the WebSocket transport is not available locally after X seconds, the dapp endpoint should display user guidance (e.g. download a wallet) and present the opportunity to connect to a remote wallet endpoint using one or more of the other association mechanisms.
+If the WebSocket transport is not available locally after no less than 30 seconds, the dapp endpoint should display user guidance (e.g. download a wallet) and optionally present the opportunity to connect to a remote wallet endpoint using one or more of the other association mechanisms.
```
Comment (sdlaver) [inline spec/spec.md:135]: Rationale for 30s: time for a dapp to send an association Intent, the user to pick a wallet app from the disambiguation dialog, the wallet app to start, plus some margin.

File: spec/spec.md [hunk 2]
```diff
@@ -132,11 +132,11 @@ where:
...
 
 #### Android
 
-If a wallet endpoint is installed which has registered an Activity for this URI scheme and format, it will be launched. Upon launch via this URI, the wallet endpoint should start a WebSocket server on port `port_number` and begin listening for connections to `/solana-wallet` for X seconds. This websocket server should only accept connections from the localhost.
+If a wallet endpoint is installed which has registered an Activity for this URI scheme and format, it will be launched. Upon launch via this URI, the wallet endpoint should start a WebSocket server on port `port_number` and begin listening for connections to `/solana-wallet` for no less than 10 seconds. This websocket server should only accept connections from the localhost.
```
Comment (sdlaver) [inline spec/spec.md:139]: Rationale for 10s: Chrome will slow the rate of timer-based wakeups while a web page is in the background. This delay is approx twice as long as the longest-observed dapp timer wakeup delay.

File: spec/spec.md [hunk 3]
```diff
@@ -164,7 +164,7 @@ where:
 
 This URI should be provided to the wallet endpoint through an out-of-band mechanism, detailed in the subsections below. Each of the dapp and wallet endpoints should attempt to connect to the WebSocket address `wss://<host_authority>/reflect?id=<reflector_unique_id>`. On connection, each endpoint should wait for the [Reflector protocol](#reflector-protocol) to signal that the counterparty endpoint has connected.
 
-The endpoints will each wait up to X seconds for reflection to commence. If it does not commence, the endpoints will disconnect and present appropriate error messages to the user.
+The dapp endpoint must wait no less than 30 seconds for reflection to commence. The wallet endpoint must wait no less than 10 seconds for reflection to commence. If it does not commence, the endpoints will disconnect and present appropriate error messages to the user.
```
Comment (sdlaver) [inline spec/spec.md:167]: Rationale for 30s and 10s same as above.

File: spec/spec.md [hunk 4]
```diff
@@ -225,7 +225,7 @@ The `HELLO_REQ` message is the first message sent after a connection is establis
 
 On receipt, the wallet endpoint should verify the signature of `Qd` using the association token. If signature verification is successful, the wallet endpoint should prepare and send a `HELLO_RSP` message to the dapp endpoint.
 
-If qd signature verification fails, if no `HELLO_REQ` message is received by the wallet endpoint within X seconds, or if a second `HELLO_REQ` message is received by the wallet endpoint at any time during the connection, all ephemeral key materials should be discarded, and the connection should be closed.
+If qd signature verification fails, if no `HELLO_REQ` message is received by the wallet endpoint within no less than 10 seconds, or if a second `HELLO_REQ` message is received by the wallet endpoint at any time during the connection, all ephemeral key materials should be discarded, and the connection should be closed.
```
Comment (sdlaver) [inline spec/spec.md:228]: Rationale for 10s same as above.

File: spec/spec.md [hunk 5]
```diff
@@ -255,7 +255,7 @@ Upon sending of the `HELLO_RSP` message by the wallet endpoint, and receipt of t
 
 Once each endpoint has calculated the ephemeral shared secret, they should proceed to providing or consuming the [Wallet RPC interface](#wallet-rpc-interface).
 
-If either public keypoint `Qd` or `Qw` is not valid, if no `HELLO_RSP` message is received by the dapp endpoint within X seconds, or if a second `HELLO_RSP` message is received by the dapp endpoint at any time during the connection, all ephemeral key materials should be discarded, and the connection should be closed.
+If either public keypoint `Qd` or `Qw` is not valid, if no `HELLO_RSP` message is received by the dapp endpoint within no less than 10 seconds, or if a second `HELLO_RSP` message is received by the dapp endpoint at any time during the connection, all ephemeral key materials should be discarded, and the connection should be closed.
```
Comment (sdlaver) [inline spec/spec.md:258]: Rationale for 10s same as above.

File: spec/spec.md [hunk 6]
```diff
@@ -957,7 +957,7 @@ On a disconnection:
 - If the connection is part of the fully open reflections data set, the entry will be removed and the other connection closed as well
 - Otherwise, the entry for the connection will be removed from the half-open reflections data set
 
-Entries in the half open data set should be removed, and the connection closed, if still present in this set  X seconds after being added. Entries in the fully open data set should be removed, and both connections closed, if still present in this list X seconds after being added.
+Entries in the half open data set should be removed, and the connection closed, if still present in this set no less than 30 seconds after being added. Entries in the fully open data set should be removed, and both connections closed, if still present in this list no less than 90 seconds after being added.
```
Comment (sdlaver) [inline spec/spec.md:960]: Rationale for 30s same as above.

Rationale for 90s - no specific rationale, but this puts a 2 min cap (30s + 90s) on reflector connections, protecting the server from non-malicious resource exhaustion.

File: spec/spec.md (hunk 1 above)
Comment (steveluscher) [reply spec/spec.md:135]: Such double, much negative, very words, wow.



Maybe “If a dapp's requests to the local WebSocket address continue to go unanswered after 30 or more seconds, it should…”

¯\\\_(°ペ)\_/¯

**Review by steveluscher — APPROVED** [review]

File: spec/spec.md (hunk 1 above)
Comment (sdlaver) [reply spec/spec.md:135]: Perhaps "at least"? I was trying to keep the same wording throughout the spec, while at the same time expressing that these are floor values, not mandated exact timeouts.

Comment (steveluscher) [reply spec/spec.md:135]: It's not a big deal. Either or.

**Pull request merged**